### ISSUE TITLE
MC cp size is bigger than real size

### ISSUE THREAD

**Issue by kybom** [minio/mc#2297]
## Expected behaviour
MC cp size is equal to real zise.
## Actual behaviour
mc cp size(1.6G) is bigger than real zise(1.59G).

> 1.60 GB / 1.59 GB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.52% 412.63 KB/s 1h7m45s

## mc version
- (paste output of `mc version`)
Version: DEVELOPMENT.GOGET
Release-tag: DEVELOPMENT.GOGET
Commit-id: DEVELOPMENT.GOGE

## System information
centos 6.5

![qq 20171026100315](https://user-images.githubusercontent.com/24692485/32031661-34f52488-ba35-11e7-800e-a2061721ec28.png)

**Comment by vadmeste** [comment]
I think this happens when mc retries a failed request in the middle of the operation.

**Comment by ebozduman** [comment]
Could you please tell us what exact command you've used when you hit this issue and was the copy done on a single file?
I am assuming you've done a simple copy on a single file like; 
```
$ mc cp /tmp/testfile /var/tmp/testfile_copy
```
but I am having hard time reproducing it.

**Comment by ebozduman** [comment]
@kybom ,
I spent quite some time on reproducing this issue by initiating network disruption (turning wifi off/on) on the server side, while `mc cp` was going on on the client side. When the network came back on, the hung copy process started continuing and the attempt failed to reproduce the issue. So, copy process completed without any problem and displayed the sizes correctly. I did repeat this test with a single 1.59GB file and with multiple files (100MB, 150MB, and 200MB) without any issues.
Code review also did not reveal any obvious/possible bugs in the code.

So, at this point, we need your help on reproducing the problem.
Please provide us more details on the reproduction; whether this issue is still reproducible/repeatable on your system/setup, what the exact size of the file in bytes is, how many files involved in the copy process are. Basically we need any information that is relevant or that might be relevant to this issue, for us to keep looking into it.
Thanks...

**Comment by kybom** [comment]
@ebozduman This issue is still reproducible on my system. The file size is 1708801009 Bytes. One file.

**Comment by kybom** [comment]
@ebozduman This issue is still reproducible on my system. The file size is 1708801009 Bytes. One file.

**Comment by kybom** [comment]
Just a copy `mc cp kk_20171021.tar.gz minio/mybucket/`

**Comment by harshavardhana** [comment]
Which version of mc are you using ?. Can you use the latest and let us know if it's reproducible ?

**Comment by kybom** [comment]
how to solve the issue 

> mc: <ERROR> Failed to copy `xxxxx_20171021.tar.gz`. Post https://www.xxxxxx.net:9000/xxxxx/xxxxx_20171021.tar.gz?uploads=: net/http: invalid header field value "unconfined_u:object_r:user_home_t:s0\x00" for key X-Amz-Meta-Security.selinux

**Comment by harshavardhana** [comment]
This is expected the object storage you are using uses incompatible http header values on objects. This is part of Go http client we do not have control on this. So we cannot fix it.

The object storage you are using is implemented incorrectly in following HTTP spec.

https://github.com/golang/net/blob/master/lex/httplex/httplex.go#L270

You should ask your object storage vendor to fix this properly and follow http spec.

**Comment by harshavardhana** [comment]
Btw is your server by any chance minio ?

**Comment by harshavardhana** [comment]
```js
const Async = require('async')
const AWS = require('aws-sdk')
const UUID = require('node-uuid')

var options = {
    accessKey: 'USWUXHGYZQYFYFFIT3RE',
    secretKey: 'MOJRH0mkL1IPauahWITSVvyDrQbEEIwljvmxdq03',
    region: 'us-east-1',
    endpoint: 'http://localhost:9000'
};

let s3opts = {
    signatureVersion: 'v4',
    accessKeyId: options.accessKey,
    secretAccessKey: options.secretKey,
    region: options.region,
}

s3opts = Object.assign(s3opts, {
    endpoint: options.endpoint,
    s3ForcePathStyle: true,
    s3BucketEndpoint: false,
    sslEnabled: false,
})

var s3Client = new AWS.S3(s3opts);
var bucket = "testbucket";

s3Client.putObject({
    Bucket: bucket,
    Key: "my-objectname",
    Metadata: {"security.selinux": "unconfined_u:object_r:user_home_t:s0\x00"},
}, (err, data) => {
    if (err) {
        console.log(err);
    }
    console.log(data);
})
```

This example throws an error 

```
{ NetworkingError: The header content contains invalid characters
    at ClientRequest.setHeader (_http_outgoing.js:370:11)
    at new ClientRequest (_http_client.js:131:14)
    at Object.request (http.js:26:10)
    at features.constructor.handleRequest (/home/harsha/aws-sdk/js/node_modules/aws-sdk/lib/http/node.js:43:23)
    at executeSend (/home/harsha/aws-sdk/js/node_modules/aws-sdk/lib/event_listeners.js:241:29)
    at Request.SEND (/home/harsha/aws-sdk/js/node_modules/aws-sdk/lib/event_listeners.js:256:9)
    at Request.callListeners (/home/harsha/aws-sdk/js/node_modules/aws-sdk/lib/sequential_executor.js:101:18)
    at Request.emit (/home/harsha/aws-sdk/js/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/home/harsha/aws-sdk/js/node_modules/aws-sdk/lib/request.js:668:14)
    at Request.transition (/home/harsha/aws-sdk/js/node_modules/aws-sdk/lib/request.js:22:10)
  message: 'The header content contains invalid characters',
  code: 'NetworkingError',
  region: 'us-east-1',
  hostname: 'localhost',
  retryable: true,
  time: 2017-11-04T21:06:53.361Z }
```

This is the expected error i would see on AWS SDKs, Minio SDKs - i am not sure how you created an object with the header value which is incorrect according to HTTP spec.

**Comment by kybom** [comment]
The release mc  version does not the net/HTTP error. The lastest commit has the  error"net/http...".
Minio server is mine, "VERSION:
  2017-09-29T19:16:56Z".

**Comment by harshavardhana** [comment]
> The release mc version does not the net/HTTP error. The lastest commit has the error"net/http...".
> Minio server is mine, "VERSION:
> 2017-09-29T19:16:56Z".

I do not follow anything that is being said here can you open a new issue and close this one? @kybom

**Comment by ebozduman** [comment]
Hey @kybom ,

Sorry, but we are kind of confused here.

Please clarify the followings:

1. You still see the original size mismatch problem after running `mc cp` command this issue is filed for and the version of the mc you see it with is what? Released mc or the latest mc? You wrote in your original comment:
```
Version: DEVELOPMENT.GOGET
Release-tag: DEVELOPMENT.GOGET
Commit-id: DEVELOPMENT.GOGE
```
which tells us you've used the dev code, but now you say `The release mc version does not (have) the net/HTTP error.`, which implies us that you've used `released mc version` when you saw the size issue. 

Please clarify which one it is. If it is the release version, which mc release was it? 

By any chance, if the original version info is incorrect/wrong, please modify and correct it.

2. It looks like you see a separate problem, net/HTTP error, with the latest mc code, right? Please file a separate issue for it and just make a reference here in this issue.

3. We assume you've used the same minio server, `"VERSION: 2017-09-29T19:16:56Z"` in both cases/scenarios. Please confirm if this is correct.

Thanks for all your help...

**Comment by kybom** [comment]
Back to the size erros.I download mc  from [https://dl.minio.io/client/mc/release/linux-amd64/mc](url).
the version has the size error.
My minio version "VERSION: 2017-09-29T19:16:56Z".

**Comment by harshavardhana** [comment]
We are not able to reproduce this on our end @kybom

**Comment by ebozduman** [comment]
@kybom , unless you have more information, we close this issue since we couldn't reproduce it and make any progress on it.
Please reopen this issue or open up a new one if you can reproduce it with a different file.

**Comment by harshavardhana** [comment]
Re-opening as I was able to reproduce a behavior where `mc cp` would progress further than the Total amount , this happens upon a retry request progress bar keeps moving forward. 

```
/etc/hosts:                  896 B / 224 B ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 400.00% 549 B/s -1s^C
```

**Comment by harshavardhana** [comment]
Sent the PR https://github.com/minio/minio-go/pull/1070 we need also `mc` change for this.

**Comment by harshavardhana** [comment]
This is fixed already  minio/minio-go#1070
